File: 0013-roman-to-integer/solution.java
class Solution {
    public int romanToInt(String s) {
        int res = 0;
        int max = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            int value = getIntFromCharRoman(s.charAt(i));
            if (value >= max) {
                max = value;
                res += value;
            } else
                res -= value;
        }
        return res;
    }
    
    public static int getIntFromCharRoman(char c) {
        int value = 0;
        switch (c) {
            case 'I':
                value = 1;
                break;
            case 'V':
                value = 5;
                break;
            case 'X':
                value = 10;
                break;
            case 'L':
                value = 50;
                break;
            case 'C':
                value = 100;
                break;
            case 'D':
                value = 500;
                break;
            case 'M':
                value = 1000;
                break;
            default:
                break;
        }
        return value;
    }
}

File: 0013-roman-to-integer/solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        table = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000,
        }

        i = 0
        n = len(s)
        res = 0

        # IX => 9
        # VII => 8
        # IV => 4
        while i < n:
            c = s[i]
            if i < n - 1:
                if table[c] < table[s[i + 1]]:
                    res += table[s[i + 1]] - table[c]
                    i += 2
                    continue

            res += table[c]
            i += 1
        return res

